### PR TITLE
Add github actions CI config, improve tests, fix few typos and hlint warnings

### PR DESCRIPTION
**Pull request by jhrcek**
Playing with this package locally I found out that there are couple of tests failing on master.

It was few tests that expect ConversionFailed exception which wasn't thrown.
I improved the tests so that they don't just except any old exception to be thrown, but a particular exception constructor with particular messages and to also PRINT the exception being expected if nothing is thrown.

I added CI config so that this kind of failure can be caught before merging changes

TODO
- [x] Remove travis config as nowadays people mostly use github actions
- [x]  Add github actions config auto-generated by [haskell-ci](https://github.com/haskell-CI/haskell-ci/)
- [x] Fix couple of typos
- [x] Fix some hlint warnings
- [x] decide what to do about the current (maybe not a bug) behavior in commented out tests.

### CODE REVIEW

File: test/Errors.hs [hunk 1]
```diff
@@ -56,54 +52,83 @@ testErrorsColumns TestEnv{..} = TestCase $ do
...
+  assertResultErrorThrown (query_ conn "SELECT 'true'" :: IO [Only Bool])
+    ConversionFailed {errSQLType = "TEXT", errHaskellType = "Bool", errMessage = "expecting an SQLInteger column type"}
+  assertResultErrorThrown (query_ conn "SELECT 2" :: IO [Only Bool])
+    ConversionFailed {errSQLType = "INTEGER", errHaskellType = "Bool", errMessage = "bool must be 0 or 1, got 2"}
   -- Mismatching types (sources don't match destination string types (text, string)
-  assertResultErrorCaught (query_ conn "SELECT 1" :: IO [Only T.Text])
-  assertResultErrorCaught (query_ conn "SELECT 1" :: IO [Only LT.Text])
-  assertResultErrorCaught (query_ conn "SELECT 1.0" :: IO [Only T.Text])
-  assertResultErrorCaught (query_ conn "SELECT 1.0" :: IO [Only LT.Text])
+  {- TODO these 4 tests are NOT actually throwing any error
```
Comment (jhrcek) [inline test/Errors.hs:89]: I commented out these 4 tests, because they are not actually throwing any error - they correctly parse the results as "1" and "1.0".

This is also happening on master (it's not bug introduced by changes in this PR). Should this test be removed or should the library behavior be changed so that errors are thrown?

File: test/Errors.hs [hunk 2]
```diff
@@ -23,15 +23,11 @@ import           Data.Time (Day, UTCTime)
...
--- the contents of error messages.  Another option would be to log
--- them (would be useful), but I don't know if HUnit has any logging
--- mechanisms.  Just printing them as is will look like the tests are
--- hitting errors and would be confusing.
-assertResultErrorCaught :: IO a -> Assertion
-assertResultErrorCaught action = do
-  catch (action >> return False) (\(e :: ResultError) -> length (show e) `seq` return True) >>=
-    assertBool "assertResultError exc"
+assertResultErrorThrown :: IO a -> ResultError -> Assertion
+assertResultErrorThrown action expectedError =
```
Comment (jhrcek) [inline test/Errors.hs:27]: This is I think better way to test exception throwing behavior - actually compare the expected exception with what's being thrown.

File: sqlite-simple.cabal [hunk 3]
```diff
@@ -23,7 +23,7 @@ Category:            Database
 Build-type:          Simple
 
 Cabal-version:       >= 1.10
-
+tested-with:         GHC == 8.8.4 || == 8.10.7 || == 9.0.2 || == 9.2.8 || == 9.4.8 || == 9.6.3 || == 9.8.1
```
Comment (jhrcek) [inline sqlite-simple.cabal:26]: You can check in my repo that CI is passing for all these ghc versions (https://github.com/jhrcek/sqlite-simple/actions/runs/7472016831/job/20333355139?pr=10). The CI would start working in this repo after this PR is merged.

File: test/Errors.hs (hunk 1 above)
Comment (jchia) [reply test/Errors.hs:89]: I have reverted the commit responsible for the test failure. (See #98 & #99). Did you make these test changes just for the failure or were you also trying to improve the test or test code in other ways?

**Comment by jhrcek** [comment]
Hey @jchia 
thanks for resolving the test issue.

To give you some context here's where this PR is coming from.
- At first I just wanted to read this package source code to get some inspiration for similar functionality I need to implement in a private package.
- I noticed few typos and easy-to-fix hlint warnings that I fixed
- I executed the tests locally to confirm I didn't break anything, but then I got the test failure. It wasn't clear to me why the tests were failing.
- Found out the "TODO" in tests: https://github.com/nurpax/sqlite-simple/pull/110/files#diff-9d86bde5ab0508a8752895f7a2185ccbb625ada7b0712bbc769ea6351a2580c8L26 and decide to use more reasonable and "better assertion error giving" helper function - IMO it's good not to just check that any old exception was thrown, but also to verify that the exception thrown is exactly the one we were expecting AND to show the expected exception if it isn't thrown at all.
- Then I figured it would be great to have CI config to prevent merging test-breaking changes to master (we depend on this package in production and it would be great to catch issues like this before the package is released). I find haskell-ci really useful and low-maintenance effort way to do that (you just need to add new GHC version to tested-with section in cabal and rerun `haskell-ci regenerate` with sufficiently recent version of `haskell-ci`).

File: Database/SQLite/Simple/FromField.hs [hunk 4]
```diff
@@ -1,6 +1,7 @@
-{-# LANGUAGE CPP, DeriveDataTypeable, DeriveFunctor  #-}
-{-# LANGUAGE FlexibleInstances, TypeSynonymInstances #-}
-{-# LANGUAGE ScopedTypeVariables      #-}
+
```
Comment (jchia) [inline Database/SQLite/Simple/FromField.hs:1]: Extra newline. Maybe you inherited it from the recently reverted commit.

File: test/Errors.hs [hunk 5]
```diff
@@ -306,4 +306,4 @@ testErrorsSavepoint TestEnv{..} = TestCase $ do
           return True
         [] ->
           return False
-        _ -> error "should have only one row"
\ No newline at end of file
+        _ -> error "should have only one row"
```
Comment (jchia) [inline test/Errors.hs]: This change belongs in 8aa062a7.

File: Database/SQLite/Simple/FromField.hs [hunk 6]
```diff
@@ -1,7 +1,9 @@
 
-{-# LANGUAGE CPP, DeriveDataTypeable #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DeriveDataTypeable #-}
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeSynonymInstances #-}
```
Comment (jchia) [inline Database/SQLite/Simple/FromField.hs:5]: This is extra.

File: README.markdown
Comment (jchia) [inline README.markdown:1]: I think the change to this file does not belong in this commit. The change is not about improving tests. It's mainly about switching from Travis to Github. It makes more sense in 8b80afcb.

File: Database/SQLite/Simple/FromField.hs (hunk 6 above)
Comment (jhrcek) [reply Database/SQLite/Simple/FromField.hs]: Not sure what you mean.
My change removed some of the language extensions that this module was not actully using, and then formatted the rest over one line.

Or do you mean I should remove all unused language extensions (TypeSynonymInstances is not used in this module)? If so, I removed all unused extensions from all modules in the next commit.

File: Database/SQLite/Simple/FromField.hs (hunk 4 above)
Comment (jhrcek) [reply Database/SQLite/Simple/FromField.hs]: OK, removed.

**Comment by jchia** [comment]
All my content comments are in. After proper organization of the commits for a clean git history (involving force-push to this PR, no doubt), this PR should be good to go.

Specifically, a commit history where changes go in the right commits, with no internal mistakes that are corrected in separate commits.

**Comment by jhrcek** [comment]
Ok, I reorganized the commit history so it's now hopefully more coherent. Please check again.

**Comment by jchia** [comment]
Thanks for the contribution.
